src/mysql.erl

```diff
@@ -712,22 +712,31 @@ encode(Conn, Term) ->
 
 %% @doc Executes the given queries and prepares the given statements after a
 %% connection has been made.
+%%
+%% If any of the queries or prepares fails, the connection is closed and an
+%% exception is raised.
 -spec execute_after_connect(connection(), [iodata()], [{atom(), iodata()}])
     -> ok.
 execute_after_connect(Conn, Queries, Prepares) ->
-    lists:foreach(fun (Query) ->
-                      case query(Conn, Query) of
-                          ok -> ok;
-                          {ok, _} -> ok;
-                          {ok, _, _} -> ok
-                      end
-                  end,
-                  Queries),
-    lists:foreach(fun ({Name, Stmt}) ->
-                      {ok, Name} = prepare(Conn, Name, Stmt)
-                  end,
-                  Prepares),
-    ok.
+    try
+        lists:foreach(fun (Query) ->
+                          case query(Conn, Query) of
+                              ok -> ok;
+                              {ok, _} -> ok;
+                              {ok, _, _} -> ok
+                          end
+                      end,
+                      Queries),
+        lists:foreach(fun ({Name, Stmt}) ->
+                          {ok, Name} = prepare(Conn, Name, Stmt)
+                      end,
+                      Prepares),
+        ok
+    catch
+        ?EXCEPTION(Class, Reason, Stacktrace) ->
+            catch stop(Conn),
+            erlang:raise(Class, Reason, ?GET_STACK(Stacktrace))
```

Comment:
Yes, exactly, this is what I was thinking. So you agree this is a good way?

Just one thing: Perhaps it would be good to use stop/2 with a timeout (any value less than infinity, perhaps just reusing `?default_connect_timeout`) to make sure this can't hang the start_link call forever. What do you think?

Reply:
Yes, I agree.  :)

A timeout is good, I'd say.

But, if it expires, it will also kill the calling process, since they are linked at this point already. But I guess expiry is very unlikely, and I'm not sure if this could be a problem even if it happens.

We could also use a short-lived process to do the `mysql:stop` call and unlink the connection process from the calling process.

Reply:
I too think expiry is very unlikely. Not worth adding the complexity of a short-lived process IMO. Adding a timeout is very simple though. It kills the calling process (unless it's trapping exits) but that's better than hanging.

Reply:
Ok, done :)